### ISSUE TITLE
Some keys don't work for macOS Sierra clients

### ISSUE THREAD

**Issue by grigorovich** [symless/synergy#5592]
Server: Synergy server v1.8.2-stable-130a248 ; MacOSX 10.11.6 / macbook air
Client: Synergy client v1.8.2-stable-130a248 ; MacOSX 10.12 GM / imac
Scancode application: Many Tricks - Key Codes v2.1

Some keys: 
CTRL + Up (Mission Control)
CTRL + Down (Application windows)
CTRL + Left (Switch to left desktop if you have additional virtual desktop)
CTRL + Right (Switch to right desktop if you have additional virtual desktop)
F11 - Show Desktop (if checked in system preferences)
FN+F3 ; FN+F4 (or just F3/F4 button like standard mac keyboard)

They does'n send by SynergyServer correctly or doesn't receive by SynergyClient correctly...
### 1) Scancode application on server - just server's keyboard:

Unicode: 63236 / 0xf704
Keys: F1
Key Code: 122 / 0x7a 
Modifiers: 8388864 / 0x800**100** **<<<<<<<<<<<<<<** 
Modifiers Description:
NX_SECONDARYFNMASK / kCGEventFlagMaskSecondaryFn / NSFunctionKeyMask
    = 8388608 / 0x800000
**NX_NONCOALSESCEDMASK / kCGEventFlagMaskNonCoalesced
    = 256 / 0x100**  **<<<<<<<<<<<<<<** That attribute is absent in **2) situation below**
### 2) Scancode application on client - receive server's Synergy keyboard scancode:

Unicode: 63236 / 0xf704
Keys: F1
Key Code: 122 / 0x7a  **<<<<<<<<<<<<<<**  OK (Same) but...
Modifiers: 8388608 / 0x800**000** **<<<<<<<<<<<<<<** ?? Not same
Modifiers Description:
NX_SECONDARYFNMASK / kCGEventFlagMaskSecondaryFn / NSFunctionKeyMask
    = 8388608 / 0x800000
                  **<<<<<<<<<<<<<<** ?? Absent/Where?
### 3) Scancode application on client - just client's keyboard:

Unicode: 63236 / 0xf704
Keys: F1
Key Code: 122 / 0x7a   **<<<<<<<<<<<<<<** OK (same)
Modifiers: 8388864 / 0x800**100**   **<<<<<<<<<<<<<<** Present. Same as 1)
Modifiers Description:
NX_SECONDARYFNMASK / kCGEventFlagMaskSecondaryFn / NSFunctionKeyMask
    = 8388608 / 0x800000
**NX_NONCOALSESCEDMASK / kCGEventFlagMaskNonCoalesced
    = 256 / 0x100**   **<<<<<<<<<<<<<<** Present

In 3rd situation result of native mac keyboard on client same as native mac keyboard on server.
But in 2nd situation synergy client did not received additional attribute like 0x100.

**Comment by Jayceh** [comment]
Just found this same behavior. Running 1.8.2, linux server, mac client. Was working, just upgraded the mac to sierra and now do not get ctrl to work in any hotkey sequence.  Ctrl does work when mouse-clicking though.  Something specifically when upgrading to Sierra.

**Comment by neybar** [comment]
I'm having this same problem.  With linux 1.8.2 server and mac Sierra client, the Ctrl key doesn't seem to register.

**Comment by dugger** [comment]
Same problem, but I've noticed that Ctrl + (0-9) still works.

**Comment by Kenblair1226** [comment]
Same here, Ctrl + UP, DOWN, LEFT, RIGHT do not seem working.

**Comment by vmtyler** [comment]
Same here, Sierra Client, El Capitan Server. 1.8.3-rc2-89c9c58

**Comment by alanhoyle** [comment]
I am having trouble with cmd+left/right/up/down.  Note that the Keyboard viewer does, in fact, display that I'm pressing cmd-left (e.g.), but the UI doesn't respond.

**Comment by Jayceh** [comment]
As an added note, on a different set of machines, 1.8.2 across the board, Sierra server to Linux client, the keys work fine, but with Sierra server and Sierra client, it also doesn't work. So very specifically a Sierra *client* issue.

**Comment by nlyan** [comment]
Thanks guys. This is obviously a regression and will be a priority for 1.8.4

**Comment by alanhoyle** [comment]
Weird thing:  the keyboard viewer switches from a 105 key layout (if I hit a key on the 105 key external attached to my mac client) to a 77 key layout if I hit a key on the linux server keyboard.

**Comment by stephenotalora** [comment]
hi guys any updates on this? has the issue been resolved as part of v1.8.4? and if so when will it be released?

**Comment by focalintent** [comment]
From doing some digging/testing on my machines here - it appears to be less about key combinations not going through and more about certain mission control actions not getting triggered properly by synergy.  If I change the keyboard shortcuts for mission control (or pagination between desktops) to other keyboard combinations - they still get ignored.  Ditto the function key options for triggering mission control (among other things) - so I think this is less about the specific key combinations and more about sierra ignoring synergy as a source of events that trigger mission control/virtual desktop events.

**Comment by andrewr01** [comment]
@ivelrrat I agree, I haven't been able to use the arrow keys with mission control. Like @grigorovich posted initially.

But the following DO work:
Switch to Desktop 1 = ctrl + 1
Switch to Desktop 2 = ctrl + 2

I have changed the "Move left a space" and "Move right a space" shortcuts to ctrl + L and ctrl + R and they work.

Also changed the non-functional "Show Desktop" = F11 to command + D and Mission Control to ctrl + alt + M

**Comment by focalintent** [comment]
I'll try again tomorrow on my setup - I was having none
of the key combinatIons that I tried working.  (Ditto swipe gestures for switching desktops, for that matter)

**Comment by icorbett** [comment]
I went ahead and tried reverting to each stable release from 1.8.4 back through 1.7.2 and believe I have exhausted the possibility that this is a regression but instead a new behavior of Siera (10.12).

Both when _successfully_ setting Keyboard Shortcuts in System Preferences to use CTRL + Up, CTRL + Down, CTRL + Left, CTRL + Right; and _unsuccessfully_ trying to trigger those same key sequences, I see the following message in the synergy Log:
```
[2016-10-11T13:34:37] INFO: found key in group 1
```

For the record I am using synergys built directly from git tags on a Linux machine and synergyc on macOS Siera (10.12) installed from dmg images.

**Comment by cocacrave** [comment]
having same issue since macos sierra. i saw the update for 1.8.4 and got excited to see if it was fixed but i guess next version?

**Comment by dgmiller81** [comment]
All -

I am able to confirm the same behavior.   Config below:

- Windows 10 Pro (server) running [Synergy 1.8.4-stable]
- OS X Sierra 10.12 (client) running [Synergy 1.8.4-stable]

Just wanted to share for reference.   That is all.

**Comment by icorbett** [comment]
There appears to be a build of synergy-v1.8.5-stable-984ca48 today (http://symless.com/nightly?filter=1.8.5) but it appears to be a mistake, 984ca48 does not appear to be a valid commit, and it certainly doesn't fix issue #5592.

**Comment by focalintent** [comment]
Following up with my own setup here -- if I go into keyboard | shortcuts on my client mac and set the key commands for move left/move right to ctrl-cmd-[ and ctrl-cmd-] then those are working over synergy.  Apparently the initial set of alternate key commands I was trying to use were things that were still filtered. 

(Note that multi-touch gestures are still completely lost - in my setup both host and client machines are macOS sierra machines)

PS. HI @icorbett :)

**Comment by chevex** [comment]
I'm having the same issue after upgrading to Sierra. Googled and couldn't find anything because I kept searching for "OS X Spaces not working with synergy" or variations of that query. I hadn't thought to search for "synergy sierra mission control" until today and now here I am. Glad I'm not the only one as it's been driving me crazy.

I'm using Windows 10 as the host and my Macbook Pro with OS X Sierra 10.12 as the client. This began immediately after upgrading to Sierra. I've tried several older versions of Synergy client & host and they all have the same issue up through the current stable build. 

It does not matter what keybinding I choose in the mission control preferences. They are all ignored. The odd thing is that the arrow keys clearly work as I can arrow around text and other thing on my macbook using the arrow keys on the host machine. I can also use Ctrl just fine for other things like copy, paste, or killing programs/clearing the screen in OS X terminal.

One thing to note is that there is a second dropdown in mission control keybindings that has a couple mouse options for accessing mission control. I tried setting that to "Middle Mouse Button" and that actually worked over Synergy. I turned it back off because I don't particularly like using my middle mouse button for that, but it did indeed work over synergy.

![image](https://cloud.githubusercontent.com/assets/874394/19662882/b59ba81a-99f6-11e6-973c-6e576e7c5ea7.png)

**Comment by andrewr01** [comment]
![screen shot 2016-10-25 at 11 37 35 am](https://cloud.githubusercontent.com/assets/10983525/19666528/7696bc5c-9aa7-11e6-8cc8-9a96fa9de977.png)
These keys work for me over Synergy, just the original settings which are the same as your screenshot don't work for me, and any combination that uses arrow keys in the mix.

**Comment by chevex** [comment]
Interesting. I tried to use F1 and Left control and they did not work. I didn't go through the entire list though.

**Comment by chevex** [comment]
Actually, how did you choose a letter key as an option? I can't seem to be able to use anything other than the predefined choices.

**Comment by andrewr01** [comment]
System Preferences > Keyboard > Shortcuts > Mission Control

**Comment by alanhoyle** [comment]
In any web-based applications the COMMAND-left/right are getting interpreted as "back/forward" respectively and taking me to the previous/next web page when sent through Synergy.  Everything works as expected with the native keyboard.  I've tested in Chrome, Safari, Firefox, and Mailplane (a gmail browser app).

**Comment by rdamazio** [comment]
Even more interesting - if I use System Preferences > Keyboard > Shortcuts > Mission Control to assign a shortcut, and press F11 over synergy, it captures that as the shortcut! But then actually using it fails.

**Comment by MattzFiber** [comment]
Can confirm I am having this too 
CMD - Arrows are totally not working for me at all :(

**Comment by sness** [comment]
Confirmed not working for me as well

**Comment by chevex** [comment]
@andrewr01 thanks! Confirmed that works. I was at least able to do that and map those combinations to left and right tilt on my mouse wheel. At least I have a workaround. Was driving me crazy.

**Comment by cocacrave** [comment]
@chevex how did you assign left/right mouse tilt to move left/right space? could you walk me through?

**Comment by chevex** [comment]
@cocacrave oh I just have a Logitech mouse with mappable buttons. The mouse wheel on it tilts left and right and I used to map those buttons to ctrl + left arrow and ctrl + right arrow. Those keys don't work over synergy anymore but thanks to @andrewr01 I was able to go to system preferences > keyboard > shortcuts > mission control and map "ctrl + option + L" and "ctrl + option + R" to move spaces left and right and those seem to work over synergy so I was able to map the new keys to my mouse wheel tilt buttons.

**Comment by Rimbo** [comment]
Upgraded to 1.8.5. Issue remains. (Created duplicate issue 5708; that one can be closed.)

**Comment by XinyuHou** [comment]
Hi guys,
Could you test our latest 10.12 build on your sierra to see if the problem persists? Thanks.
http://symless.com/nightly?filter=synergy-v1.8.6-rc1-7ce6905

**Comment by mcg** [comment]
Still broken for me with synergy-v1.8.6-rc1-7ce6905.

**Comment by markrcote** [comment]
Still broken here as well. Both server and client running synergy-v1.8.6-rc1-7ce6905.

**Comment by vmtyler** [comment]
Same. When I let CTRL & Arrow all I get is “found key in group 0” in the log.


On November 1, 2016 at 11:43:46 AM, Mark Côté (notifications@github.com) wrote:

Still broken here as well. Both server and client running synergy-v1.8.6-rc1-7ce6905.

**Comment by Rimbo** [comment]
Same as the above. "found key in group 0," no luck. Had 1.8.6-rc1-7ce6905 on both client (10.12) and server (10.11).

**Comment by icorbett** [comment]
I also see the same behavior as before with synergy-v1.8.6-rc1-7ce6905 on both macOS Sierra (10.12.1) client and Linux server however I continue to get the following error which is slightly different from what it appears others are seeing:

`[2016-11-02T10:19:16] INFO: found key in group 1`

**Comment by gservice** [comment]
I have same issue.  I'm using v1.8.5.
For now, I add Mission Control in Dock so that at least I can access that functionality.

**Comment by solokacher** [comment]
1.8.3 and 1.8.5 client on macOS Sierra has the same problem on my side. Slash "/" also doesn't work.

**Comment by paintedballerina** [comment]
1.8.5 installed on both machines.   Sierra server | Win7 client.
Mouse works, clicking mouse works... no keystrokes.

**Comment by ihji** [comment]
+1

**Comment by Unzensiert** [comment]
+1

**Comment by moexius** [comment]
Any updates? I have tried to remap the keys arrow left/right + ctrl to something else, not even using ctrl at all, but I believe the issue is with synergy not properly tapped to application switching.

**Comment by sandinak** [comment]
Sys Eng/Coder here.. I can has the same problem.. tried to switch my ctrl char ... I've been using ctrl-arrow since fvwm2 days .. can't break the habit.  Any assistance I can give.. I am more than willing to help out..

**Comment by sandinak** [comment]
All... FWIW .. I filed a bug on this a whilst back.. i know they're on it ;)

**Comment by XinyuHou** [comment]
Hi guys, thank you for all the feedbacks. We are still investigating this issue. So far, it seems macOS Sierra does not support some keys simulation with its native Core Graphic function CGEventPost . Those keys are Mission Control, Launchpad  and Option-Command-Escape. We are trying to find if there is any alternative way to simulate that. Thanks for your patience.

**Comment by XinyuHou** [comment]
Hi guys, can you test this nightly for Mac? It's valid for Mac 10.9 and above.
http://symless.com/nightly?filter=synergy-issue5592-ctrl-up-key-rc1-e8145aa-MacOSX-x86_64.dmg

**Comment by andynu** [comment]
@XinyuHou: 💚 I was experiencing the F1 key not working in macOS Sierra 10.12.1. This nightly release fixed the issue in my context. Wonderful!

**Comment by gservice** [comment]
@XinyuHou I tested Mission Control functionality with key ( ctrl + up ) and switching desktop with key ( ctrl + left and right ). It works well. thank you. plz let us know if it goes into stable version. great job.

**Comment by mcg** [comment]
Ctrl+left/right is working again for me! Thanks for the fix.

**Comment by moexius** [comment]
Same here. So far so good. Mission control works again! thanks for the fix.

**Comment by dugger** [comment]
👍   Thank you!!

**Comment by AndreasDolphin** [comment]
Had the same issue with 1.8.2, the linked nightly build fixed it for me :D

**Comment by XinyuHou** [comment]
Thank you for the feedbacks, guys.

@gservice
Yeah, it will be in our next stable release which would be in 3 weeks. Please sit tight. Thanks for your patience.

**Comment by AndreasDolphin** [comment]
As a side note as I just seriously started using my work Macbook, while CMD+Up/Left/Right works through Synergy, I use those a lot, taking screenshots seems wonky. CMD+Shift+4 that's used for snapping regions only reacts on the laptop keyboard and not from my Windows host. Related?

**Comment by XinyuHou** [comment]
@AndreasDolphin 
Which version of Synergy are you using?

**Comment by AndreasDolphin** [comment]
Just checked, 1.8.3-stable-db9181b on the server side, Windows, basic. I was going to update but didn't want to close down all the required applications just yet :P Maybe on Friday. The Mac client has version 1.8.6-rc1-e8145aa which fixed the other keys.

**Comment by XinyuHou** [comment]
@AndreasDolphin 
Please update your server to the latest stable. I have tried on my Windows 10 server and Mac 10,12 client. CMD+Shift+4 works to get regional screenshot.

**Comment by AndreasDolphin** [comment]
Server is now running 1.8.5-stable-a18eba7, issue persists. Not sure if there is any log I can look at to see what's going on except for what is printed in the client window.

**Comment by XinyuHou** [comment]
@AndreasDolphin 
Please create a new issue and use our template to put as many details as possible. Some logging from both server and client would be helpful. Thanks.

**Comment by Adam2Marsh** [comment]
Hi @XinyuHou

Tested the MacOs Nightly build and the issue is not resolved for me.

I'm running Ubuntu Host 1.8.5 Stable, when pressing control left or right on the mac I still just see the found key in group 0.

**Comment by XinyuHou** [comment]
@Adam2Marsh 
Please let us know your OS and Synergy verison info. Logging from both your server and client would be helpful. Please use Debug2 log level. You can change it in Edit->settings->Logging on  Ubuntu and Synergy->Preferences->Logging

**Comment by stephenotalora** [comment]
Thank you @XinyuHou works perfect Mac to Mac! happy camper here 👍

**Comment by andrewr01** [comment]
Fixed for me with the nightly BUT ... Now when I use Parallels Client to RDP to my Windows Server, no Windows key is sent. Parallels Client is set to use CMD for Windows Key. I have reverted to 1.8.5 stable on Mac Sierra client as I need the Windows key and can configure non-default keys for switching desktops on the Mac.

**Comment by vmtyler** [comment]
Yup. using the synergy-issue5592-ctrl-up-key-rc1-e8145aa-MacOSX-x86_64 build on 10.12 client with 10.12 server and keys are working again


On November 27, 2016 at 1:56:11 PM, Jonathan (notifications@github.com) wrote:

Thank you @XinyuHou works perfect Mac to Mac! happy camper here 👍

**Comment by andrewr01** [comment]
Fixed for me with the nightly BUT ... Now when I use Parallels Client to RDP to my Windows Server, no Windows key is sent. Parallels Client is set to use CMD for Windows Key. I have reverted to 1.8.5 stable on Mac Sierra client as I need the Windows key and can configure non-default keys for switching desktops on the Mac.

**Comment by focalintent** [comment]
Another vote for this working between macOS sierra machines - now, if only the trackpad gestures would go across...

**Comment by Rimbo** [comment]
The nightly works for me as well.

**Comment by doowb** [comment]
Today `Cmd+Tab` stopped working on my client machine.
The nightly (from Nov 22, 2016) fixed this, along with any issues I had with arrow keys.

**Comment by XinyuHou** [comment]
Merged in to 1.8.6

**Comment by alanhoyle** [comment]
Just wanted to report that this is indeed fixed in 1.8.6 for me.  Thanks again!

**Comment by lparsons** [comment]
Working for me as well, thanks!

**Comment by alanhoyle** [comment]
I just wanted to report that this version also appears to have fixed the
"cmd-option-escape doesn't bring up Force Quit" issue that I've had for
years.  Thanks!


--
  -  Alan Hoyle  -  alan@alanhoyle.com  -  http://www.alanhoyle.com/  -

**Comment by EmmanVazz** [comment]
This update works for me as well. Thanks!

**Comment by DarcieF** [comment]
Newest update messed up my keys. Control + click no longer brings up fly out submenu in any apps (can't control + click on an element with Chrome for example or to view functions on a layer in Photoshop). Option is also acting strangely, can not option + click to hide open apps. This really disrupts my work flow.

**Comment by XinyuHou** [comment]
@DarcieF 
Sorry for the inconvenience. 

Could you create a new issue with reproduce steps? Then we can look into the cause. Thanks.

**Comment by intel352** [comment]
Similar issue exists for me in 1.8.6, I'm unable to take a screenshot on an OSX client from an OSX host, i.e. - CMD+SHIFT+4 to trigger a select-area screenshot. The host intercepts the command every time despite cursor focus on the client.

**Comment by wwboynton** [comment]
I'm still experiencing this problem in 1.8.8, using a Sierra host and an El Cap client. Using the Keyboard Viewer, presses to the control key simply don't register.

It seems to be the same issue of host intercept, because I usually have some key remapping on (I eventually took it off to see if that was causing the issue and it was not) and I could see the control key getting pressed when I hit my physical command key.

**Comment by grigorovich** [comment]
All keys works. Even do FN+F11/F12 for example. But have some issues with some applications:
Parallels (windows 10) - CMD (win key) doesn't
UnrealEngine, Blender - some combinations like SHIFT+LMB or CMD+LMB doesn't.

But in Finder Shift+LMB works. 

Server 1.8.8,  MacOS
Client 1.8.8, MacOS

**Comment by morinted** [comment]
@XinyuHou I'm working on an unrelated project and I think I made a discovery that might help. This is not using the lower level HID calls…

I was able to get command/control keyboard shortcuts working again by adding `kCGEventFlagMaskSecondaryFn` to the modifier event flags when the target key code was >=53 (escape) but not a modifier (55-63). Not looking for a fix myself, just saw that you guys had a similar issue. All the best!
